src/nvim/lua/executor.c

```diff
@@ -1043,12 +1234,19 @@ LuaRef nlua_ref(lua_State *lstate, int index)
   lua_pushvalue(lstate, index);
   LuaRef ref = luaL_ref(lstate, LUA_REGISTRYINDEX);
   if (ref > 0) {
-    nlua_refcount++;
+    nlua_inc_refcount(lstate);
```

Comment:
This much extra complexity for a what is supposed to be a primitive operation is a bit unfortunate. But AFAIK, references tracked by nvim is only used by editor state keeping refs to lua, so these won't be used by lua threads anyway. So for now we could just revert this and document `nlua_ref` is not to be used for threads. 

 But if we do later on, we should add a single userdata for thread-local state which can passed as a _c pointer_ to internal helper functions so there only will be one register (read) access per enter c code from lua to get this struct.